File: Pycharm Projects/OverprintSimulation/custom_spot_merge.py
#
# -----------------------------------------------------------------------
# <copyright file="custom_spot_merge.py" company="Hybrid Software Helix Ltd">
#  Copyright (C) 2025 Hybrid Software Helix Ltd. All rights reserved.
# </copyright>
# <summary>
#  This example is provided on an "as is" basis and without warranty of any kind.
#  Hybrid Software Helix Ltd. does not warrant or make any representations
#  regarding the use or results of use of this example.
# </summary>
# -----------------------------------------------------------------------
#

import math
import sys

import jawsmakoIF_python
from jawsmakoIF_python import *


def custom_spot_merge():
    # Adjust to your test files folder
    test_file_path = "TestFiles/"

    try:
        # Instantiate Mako
        mako = IJawsMako.create("")
        IJawsMako.enableAllFeatures(mako)
        factory = mako.getFactory()

        # Load the document
        pdf_input = IPDFInput.create(mako)
        doc_asm = pdf_input.open(test_file_path + "Robots Plus Process Colors.pdf")
        document = doc_asm.getDocument()

        for page_index in range(document.getNumPages()):
            fixed_page = document.getPage(page_index).getContent()

            # Page bounds & raster geometry
            bounds = FRect(0, 0, fixed_page.getWidth(), fixed_page.getHeight())
            resolution = 150.0
            pixel_width = int(round(bounds.dX / 96.0 * resolution))
            pixel_height = int(round(bounds.dY / 96.0 * resolution))

            # Colorspace must be CMYK for spot merging
            cmyk = IDOMColorSpaceDeviceCMYK.create(factory)

            # Find inks and build spot list
            inks = IRendererTransform.findInks(mako, fixed_page)
            spots = IRendererTransform.inkInfoToColorantInfo(mako, inks, cmyk)

            spot_names = CEDLVectWString()
            for i in range(spots.size()):
                spot_names.append(spots.getitem(i).name)

            num_process = cmyk.getNumComponents()  # 4
            num_spots = int(spots.size())
            num_buffers = num_process + num_spots

            # Prepare frame buffers
            buffers = []
            fb = CEDLVectCFrameBufferInfo()

            for i in range(num_buffers):
                # allocate pixel buffer for this plane
                buf = bytearray(pixel_width * pixel_height)
                buffers.append(buf)

                info = IJawsRenderer.CFrameBufferInfo()
                info.bufferOfs = 0
                info.rowStride = pixel_width
                info.pixelStride = 0
                fb.append(info)

            # Render true separations
            renderer = IJawsRenderer.create(mako)
            renderer.renderSeparationsToFrameBuffers(
                fixed_page,
                8,
                True,
                pixel_width,
                pixel_height,
                cmyk,
                buffers,
                fb,
                0,
                bounds,
                spot_names
            )

            # Create writer and image
            pair = IDOMRawImage.createWriterAndImage(
                mako,
                cmyk,
                pixel_width,
                pixel_height,
                8,
                resolution,
                resolution
            )
            frame_writer = pair.frameWriter

            # Spot components to merge
            components = CEDLVectVectFloat()
            for i in range(num_spots):
                comps = CEDLVectFloat()
                vals = spots.getitem(i).components()
                for c in range(4):
                    comps.append(vals.getitem(c))
                components.append(comps)

            # Merge each spot buffer with process values
            inv255 = 1.0 / 255.0
            scanline = bytearray(pixel_width * num_process)

            for y in range(pixel_height):
                row_start = y * pixel_width
                for x in range(pixel_width):
                    for c in range(num_process):
                        idx = x * num_process + c
                        scanline[idx] = buffers[c][row_start + x]
                        for i in range(num_spots):
                            spot_val = buffers[num_process + i][row_start + x] * inv255
                            current_val = scanline[idx] * inv255
                            new_val = 1.0 - (1.0 - components.getitem(i).getitem(c) * spot_val) * (1.0 - current_val)
                            scanline[idx] = int(new_val * 255.0 + 0.5)
                frame_writer.writeScanLine(scanline)

            frame_writer.flushData()

            # Convert to RGB and save as JPEG
            rgb = IDOMColorSpaceDeviceRGB.create(factory)
            cc = IDOMImageColorConverterFilter.create(
                factory,
                rgb,
                eRelativeColorimetric,
                eBPCDefault
            )
            filtered = IDOMFilteredImage.create(factory, pair.getDomImage(), cc)

            out_jpeg = f"output_{page_index}.jpg"
            IDOMJPEGImage.encode(mako, filtered, IOutputStream.createToFile(factory, out_jpeg))

            print(f"Wrote: {out_jpeg}")

    except MakoException as ex:
        print(f"MakoException {ex.m_errorCode}: {ex.m_msg}")
        sys.exit(1)
    except Exception as ex:
        print(f"Error: {ex}")
        sys.exit(1)


if __name__ == "__main__":
    custom_spot_merge()

File: Java Projects/OverprintSimulation/src/CustomSpotMerge.java
/*
 * -----------------------------------------------------------------------
 * <copyright file="CustomSpotMerge.java" company="Hybrid Software Helix Ltd">
 *  Copyright (C) 2025 Hybrid Software Helix Ltd. All rights reserved.
 * </copyright>
 * <summary>
 *  This example is provided on an "as is" basis and without warranty of any kind.
 *  Hybrid Software Helix Ltd. does not warrant or make any representations
 *  regarding the use or results of use of this example.
 * </summary>
 * -----------------------------------------------------------------------
 */

import java.io.*;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.*;
import com.globalgraphics.JawsMako.jawsmakoIF.*;

public class CustomSpotMerge {

    public static void main(String[] args) {
        // Adjust to your test files folder
        String testFilePath = "TestFiles/";

        try {
            IJawsMako mako = IJawsMako.create();
            IJawsMako.enableAllFeatures(mako);
            IEDLClassFactory factory = mako.getFactory();

            // Load the document
            IPDFInput pdfInput = IPDFInput.create(mako);
            IDocumentAssembly docAsm = pdfInput.open(testFilePath + "Robots Plus Process Colors.pdf");
            IDocument doc = docAsm.getDocument();

            for (int pageIndex = 0; pageIndex < doc.getNumPages(); pageIndex++) {
                IDOMFixedPage fixedPage = doc.getPage(pageIndex).getContent();

                // Page bounds & raster geometry
                FRect bounds = new FRect(0, 0, fixedPage.getWidth(), fixedPage.getHeight());
                double resolution = 150.0;
                int pixelWidth = (int) Math.round(bounds.getDX() / 96.0 * resolution);
                int pixelHeight = (int) Math.round(bounds.getDY() / 96.0 * resolution);

                // Colorspace must be CMYK for spot merging
                IDOMColorSpaceDeviceCMYK cmyk = IDOMColorSpaceDeviceCMYK.create(factory);

                // Find inks and build spot list
                CEDLVectColorantInfo spots =
                        IRendererTransform.inkInfoToColorantInfo(mako,
                                IRendererTransform.findInks(mako, fixedPage), cmyk);

                CEDLVectWString spotNames = new CEDLVectWString();
                for (int i = 0; i < spots.size(); i++)
                    spotNames.append(spots.getitem(i).getName());

                int numProcess = cmyk.getNumComponents(); // 4
                int numSpots = (int) spots.size();
                int numBuffers = numProcess + numSpots;

                // Prepare frame buffers
                ByteBuffer[] buffers = new ByteBuffer[numBuffers];
                CEDLVectCFrameBufferInfo fb = new CEDLVectCFrameBufferInfo();

                for (int i = 0; i < numBuffers; i++) {
                    ByteBuffer buf = ByteBuffer.allocateDirect(pixelWidth * pixelHeight);
                    buf.order(ByteOrder.nativeOrder());
                    buffers[i] = buf;

                    IJawsRenderer.CFrameBufferInfo info = new IJawsRenderer.CFrameBufferInfo();
                    info.setBufferOfs(0);
                    info.setRowStride(pixelWidth);  // bytes per row
                    info.setPixelStride(0);         // tightly packed
                    fb.append(info);
                }

                // Render true separations
                IJawsRenderer renderer = IJawsRenderer.create(mako);
                renderer.renderSeparationsToFrameBuffers(
                        fixedPage,
                        (short) 8,
                        true,
                        pixelWidth,
                        pixelHeight,
                        cmyk,
                        buffers,
                        fb,
                        (short) 0,
                        bounds,
                        spotNames
                );

                // Create writer and image
                var pair = IDOMRawImage.createWriterAndImage(mako, cmyk, pixelWidth, pixelHeight,
                        (short) 8, resolution, resolution);

                IImageFrameWriter frameWriter = pair.getFrameWriter();

                // Spot components to merge
                CEDLVectVectFloat components = new CEDLVectVectFloat();
                for (int i = 0; i < numSpots; i++) {
                    CEDLVectFloat comps = new CEDLVectFloat();
                    CEDLVectFloat vals = spots.getitem(i).getComponents();
                    for (int c = 0; c < 4; c++)
                        comps.append(vals.getitem(c));
                    components.append(comps);
                }

                // Merge each spot buffer with process values
                final float inv255 = 1.0f / 255.0f;
                byte[] scanline = new byte[pixelWidth * numProcess];
                for (int y = 0; y < pixelHeight; y++) {
                    int rowStart = y * pixelWidth;
                    for (int x = 0; x < pixelWidth; x++) {
                        for (int c = 0; c < numProcess; c++) {
                            int idx = x * numProcess + c;
                            scanline[idx] = buffers[c].get(rowStart + x);
                            for (int i = 0; i < numSpots; i++) {
                                float spotVal = (buffers[numProcess + i].get(rowStart + x) & 0xFF) * inv255;
                                float currentVal = (scanline[idx] & 0xFF) * inv255;
                                float newVal = 1.0f - (1.0f - components.getitem(i).getitem(c) * spotVal) * (1.0f - currentVal);
                                scanline[idx] = (byte) (newVal * 255.0f + 0.5f);
                            }
                        }
                    }
                    frameWriter.writeScanLine(scanline);
                }

                frameWriter.flushData();

                // Convert to RGB and save as JPEG
                IDOMColorSpaceDeviceRGB rgb = IDOMColorSpaceDeviceRGB.create(factory);
                IDOMImageColorConverterFilter cc =
                        IDOMImageColorConverterFilter.create(factory, rgb,
                                eRenderingIntent.eRelativeColorimetric,
                                eBlackPointCompensation.eBPCDefault);
                IDOMFilteredImage filtered = IDOMFilteredImage.create(factory, pair.getDomImage(), cc);

                String outJpeg = String.format("output_%d.jpg", pageIndex);
                IDOMJPEGImage.encode(mako, filtered,
                        IOutputStream.createToFile(factory, outJpeg));

                System.out.println("Wrote: " + outJpeg);
            }
        }
        catch (Exception ex) {
            ex.printStackTrace();
            System.exit(1);
        }
    }
}
